File: java/device-manager/src/main/java/com/cccode/devicemanager/ontology/service/OntologyFunctionService.java
package com.cccode.devicemanager.ontology.service;

import com.cccode.devicemanager.ontology.model.*;
import com.cccode.devicemanager.ontology.repository.*;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import groovy.lang.Binding;
import groovy.lang.GroovyShell;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.concurrent.*;

@Service
public class OntologyFunctionService {

    private final OntologyFunctionRepository    functionRepo;
    private final OntologyFunctionLogRepository logRepo;
    private final ObjectMapper mapper = new ObjectMapper();

    private final ExecutorService executor = Executors.newCachedThreadPool();

    public OntologyFunctionService(OntologyFunctionRepository functionRepo,
                                    OntologyFunctionLogRepository logRepo) {
        this.functionRepo = functionRepo;
        this.logRepo      = logRepo;
    }

    // ===== Function CRUD =====

    public List<OntologyFunction> listFunctions() {
        return functionRepo.findAll();
    }

    public OntologyFunction getFunction(Long id) {
        return functionRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Function not found: " + id));
    }

    public OntologyFunction createFunction(OntologyFunction fn) {
        return functionRepo.save(fn);
    }

    public OntologyFunction updateFunction(Long id, OntologyFunction patch) {
        OntologyFunction fn = getFunction(id);
        if (patch.getDisplayName() != null)   fn.setDisplayName(patch.getDisplayName());
        if (patch.getDescription() != null)   fn.setDescription(patch.getDescription());
        if (patch.getScriptContent() != null) fn.setScriptContent(patch.getScriptContent());
        if (patch.getStatus() != null)        fn.setStatus(patch.getStatus());
        if (patch.getInputSchemaJson() != null)  fn.setInputSchemaJson(patch.getInputSchemaJson());
        if (patch.getOutputSchemaJson() != null) fn.setOutputSchemaJson(patch.getOutputSchemaJson());
        if (patch.getActionTypeId() != null)  fn.setActionTypeId(patch.getActionTypeId());
        return functionRepo.save(fn);
    }

    public void deleteFunction(Long id) {
        getFunction(id);
        functionRepo.deleteById(id);
    }

    // ===== Execute Function =====

    public OntologyFunctionLog executeFunction(Long id, Map<String, Object> inputData) {
        return executeFunction(id, inputData, null);
    }

    public OntologyFunctionLog executeFunction(Long id, Map<String, Object> inputData, Long actionExecutionId) {
        OntologyFunction fn = getFunction(id);
        OntologyFunctionLog log = new OntologyFunctionLog();
        log.setFunctionId(id);
        log.setActionExecutionId(actionExecutionId);
        log.setInputDataJson(toJson(inputData));

        long startMs = System.currentTimeMillis();

        try {
            Object result = runGroovyScript(fn.getScriptContent(), inputData);
            long duration = System.currentTimeMillis() - startMs;

            String outputJson;
            if (result instanceof Map) {
                outputJson = toJson(result);
            } else {
                outputJson = toJson(Map.of("result", result != null ? result.toString() : "null"));
            }

            log.setOutputDataJson(outputJson);
            log.setStatus("SUCCESS");
            log.setDurationMs(duration);

        } catch (TimeoutException e) {
            long duration = System.currentTimeMillis() - startMs;
            log.setStatus("TIMEOUT");
            log.setErrorMessage("Script execution timed out after 30 seconds");
            log.setDurationMs(duration);

        } catch (Exception e) {
            long duration = System.currentTimeMillis() - startMs;
            log.setStatus("FAILED");
            log.setErrorMessage(e.getMessage());
            log.setDurationMs(duration);
        }

        return logRepo.save(log);
    }

    private Object runGroovyScript(String script, Map<String, Object> inputData) throws Exception {
        if (script == null || script.isBlank()) {
            throw new IllegalArgumentException("Script content is empty");
        }

        Callable<Object> task = () -> {
            Binding binding = new Binding();
            binding.setVariable("input",   inputData);
            binding.setVariable("context", Map.of("timestamp", System.currentTimeMillis()));

            // Expose individual input keys as top-level variables
            if (inputData != null) {
                for (Map.Entry<String, Object> entry : inputData.entrySet()) {
                    binding.setVariable(entry.getKey(), entry.getValue());
                }
            }

            GroovyShell shell = new GroovyShell(binding);
            return shell.evaluate(script);
        };

        Future<Object> future = executor.submit(task);
        try {
            return future.get(30, TimeUnit.SECONDS);
        } catch (TimeoutException e) {
            future.cancel(true);
            throw e;
        } catch (ExecutionException e) {
            throw new RuntimeException(e.getCause() != null ? e.getCause().getMessage() : e.getMessage(), e.getCause());
        }
    }

    // ===== Logs =====

    public List<OntologyFunctionLog> listLogs(Long functionId) {
        return logRepo.findByFunctionIdOrderByExecutedAtDesc(functionId);
    }

    private String toJson(Object obj) {
        try { return mapper.writeValueAsString(obj); }
        catch (Exception e) { return "{}"; }
    }
}

File: backend/app/services/ontology_function_service.py
import json
import time
from sqlalchemy.orm import Session
from app.models.ontology import OntologyFunction, OntologyFunctionLog
from app.utils.script_executor import execute_python_script, ScriptTimeoutError


# ===== Function CRUD =====

def list_functions(db: Session):
    return db.query(OntologyFunction).all()


def get_function(db: Session, id: int) -> OntologyFunction:
    fn = db.get(OntologyFunction, id)
    if not fn:
        raise ValueError(f"Function not found: {id}")
    return fn


def create_function(db: Session, data: dict) -> OntologyFunction:
    fn = OntologyFunction(**data)
    db.add(fn)
    db.commit()
    db.refresh(fn)
    return fn


def update_function(db: Session, id: int, data: dict) -> OntologyFunction:
    fn = get_function(db, id)
    for key, value in data.items():
        if value is not None:
            setattr(fn, key, value)
    db.commit()
    db.refresh(fn)
    return fn


def delete_function(db: Session, id: int):
    fn = get_function(db, id)
    db.delete(fn)
    db.commit()


# ===== Execute Function =====

def execute_function(db: Session, id: int, input_data: dict,
                     action_execution_id: int = None) -> OntologyFunctionLog:
    fn = get_function(db, id)
    log = OntologyFunctionLog(
        function_id=id,
        action_execution_id=action_execution_id,
        input_data_json=_to_json(input_data),
    )

    start_ms = time.time() * 1000

    try:
        result = execute_python_script(fn.script_content, input_data)
        duration = int(time.time() * 1000 - start_ms)

        if isinstance(result, dict):
            output_json = _to_json(result)
        else:
            output_json = _to_json({"result": str(result) if result is not None else "null"})

        log.output_data_json = output_json
        log.status = "SUCCESS"
        log.duration_ms = duration

    except ScriptTimeoutError:
        duration = int(time.time() * 1000 - start_ms)
        log.status = "TIMEOUT"
        log.error_message = "Script execution timed out after 30 seconds"
        log.duration_ms = duration

    except Exception as e:
        duration = int(time.time() * 1000 - start_ms)
        log.status = "FAILED"
        log.error_message = str(e)
        log.duration_ms = duration

    db.add(log)
    db.commit()
    db.refresh(log)
    return log


# ===== Logs =====

def list_logs(db: Session, function_id: int):
    return db.query(OntologyFunctionLog).filter(
        OntologyFunctionLog.function_id == function_id
    ).order_by(OntologyFunctionLog.executed_at.desc()).all()


def _to_json(obj) -> str:
    try:
        return json.dumps(obj, ensure_ascii=False)
    except Exception:
        return "{}"
